### PR TITLE
Extend type checking to all float datatypes

### PR DESCRIPTION
**Pull request by CeliaBenquet**
This allows for someone to use discrete labels in a continuous way; this might change the solver to behaviors you might not be expecting so please use with caution.

fix https://github.com/AdaptiveMotorControlLab/CEBRA-dev/pull/791
fix https://github.com/AdaptiveMotorControlLab/CEBRA-dev/issues/790

### CODE REVIEW

File: cebra/data/datasets.py [hunk 1]
```diff
@@ -76,23 +76,35 @@ def __init__(
...
             )
         self.offset = offset
 
-    def _to_tensor(self, array, check_dtype=None):
+    def _to_tensor(self, array, check_dtype: str = None):
+        """Convert :py:func:`numpy.array` to :py:class:`torch.Tensor` if necessary and check the dtype.
+
+        Args:
+            array: Array to check.
+            check_dtype (list, optional): If not `None`, list of dtypes to which the values in `array`
```
Comment (stes) [inline cebra/data/datasets.py:93]: remove type here and add type annotation to function!

File: cebra/data/datasets.py [hunk 2]
```diff
@@ -76,23 +76,35 @@ def __init__(
...
         if array is None:
             return None
         if isinstance(array, np.ndarray):
             array = torch.from_numpy(array)
         if check_dtype is not None:
-            if not isinstance(array, check_dtype):
-                raise TypeError(f"{type(array)} instead of {check_dtype}.")
+            if (check_dtype == "integer" and not cebra.helper._is_integer(array)
+                ) or (check_dtype == "float" and not cebra.helper._is_floating(array)
+                ) or (check_dtype == "float_integer" and not cebra.helper._is_floating_or_integer(array)):
```
Comment (stes) [inline cebra/data/datasets.py:106]: these options should appear as `Literal["integer", "float", "float_integer"]` in the signature

**Review by stes — CHANGES_REQUESTED** [review]

File: cebra/data/datasets.py (hunk 2 above)
Comment (CeliaBenquet) [reply cebra/data/datasets.py]: done

**Comment by CeliaBenquet** [comment]
docs git action doesn't pass is fixed in #176

**Comment by stes** [comment]
docs should build again now, main tests not yet. Working on that now.

**Comment by stes** [comment]
All right, after #184 this should hopefully finally build!

**Comment by stes** [comment]
Ok, raises an actual error:
image

@CeliaBenquet wanna have a shot at fixing this? otherwise I can have a look.

File: cebra/data/datasets.py [hunk 3]
```diff
@@ -74,23 +74,40 @@ def __init__(self,
                  offset: int = 1,
                  device: str = "cpu"):
         super().__init__(device=device)
-        self.neural = self._to_tensor(neural, torch.FloatTensor).float()
-        self.continuous = self._to_tensor(continuous, torch.FloatTensor)
-        self.discrete = self._to_tensor(discrete, torch.LongTensor)
+        self.neural = self._to_tensor(neural, check_dtype="float").float()
+        self.continuous = self._to_tensor(continuous,
+                                          check_dtype="float").float()
```
Comment (stes) [inline cebra/data/datasets.py:79]:
```suggestion
        self.continuous = self._to_tensor(continuous,
                                          check_dtype="float")
```

File: cebra/data/datasets.py [hunk 4]
```diff
@@ -74,23 +74,40 @@ def __init__(self,
...
             return None
         if isinstance(array, np.ndarray):
             array = torch.from_numpy(array)
         if check_dtype is not None:
-            if not isinstance(array, check_dtype):
-                raise TypeError(f"{type(array)} instead of {check_dtype}.")
+            if (check_dtype == "int" and not cebra.helper._is_integer(array)
+               ) or (check_dtype == "float" and
+                     not cebra.helper._is_floating(array)):
+                raise TypeError(f"{array.dtype} instead of {check_dtype}.")
```
Comment (stes) [inline cebra/data/datasets.py:110]:
```suggestion
                raise TypeError(f"Array has type {array.dtype} instead of {check_dtype}.")
```

**Review by stes — COMMENTED** [review]
Fix type checking and error message

**Comment by stes** [comment]
Let's see if this fixes the error

**Review by stes — APPROVED** [review]
lgtm now

**Review by icarosadero — APPROVED** [review]
@stes I think this does the trick 👍

**Pull request merged**